mllib/src/main/scala/org/apache/spark/mllib/stat/distribution/MultivariateGaussian.scala

```diff
@@ -126,11 +120,11 @@ class MultivariateGaussian @Since("1.3.0") (
       // log(pseudo-determinant) is sum of the logs of all non-zero singular values
       val logPseudoDetSigma = d.activeValuesIterator.filter(_ > tol).map(math.log).sum
 
-      // calculate the root-pseudo-inverse of the diagonal matrix of singular values
-      // by inverting the square root of all non-zero values
-      val pinvS = diag(new DBV(d.map(v => if (v > tol) math.sqrt(1.0 / v) else 0.0).toArray))
+      // calculate the pseudo-inverse of the diagonal matrix of singular values
+      // by inverting the non-zero values
+      val pinvS = diag(new DBV(d.map(v => if (v > tol) 1.0 / v else 0.0).toArray))
 
-      (pinvS * u, -0.5 * (mu.size * math.log(2.0 * math.Pi) + logPseudoDetSigma))
+      (u * pinvS * u.t, -0.5 * (mu.size * math.log(2.0 * math.Pi) + logPseudoDetSigma))
```

Comment:
Yes that's equivalent numerically, and would require the other changes above. I'm not clear why it's better to do it this way though? it takes longer to take the square root of the eigenvalues, and then they're just multiplied back together. It's the same number of operations here and above otherwise.

I think the evidence that it's not accurate enough is the case in the JIRA and tests here, and also the Pyspark test that is wrong at the moment.